source/gui/accPropServer.py

```diff
@@ -90,24 +103,58 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
 			multiple accessible elements, it is still bound to one wx.Control.
-		@type pIDString: str
 		@param dwIDStringLen: Specifies the length of the identity string specified by the pIDString parameter.
-		@type dwIDStringLen: int
-		@param idProp: Specifies a GUID indicating the desired property.
-		@type idProp: One of the oleacc.PROPID_* GUIDS
-		@return A tuple of the out params for the `IAccPropServer::GetPropValue` method: `VARIANT* pvarValue` and `BOOL*
-		pfHasProp`. When the pfHasProp part is FALSE / self.DOES_NOT_HAVE_PROP, then the pvarValue part must be VT_EMPTY.
-		Consider using self.NO_RETURN_VALUE instead. Returning (VT_EMPTY, HAS_PROP) IS valid, meaning the property exists
-		but is empty.
+		@param idProp: Specifies a GUID indicating the desired property. One of the values from oleacc.PROPID_*
+		@return Use L{self._hasProp} to return correct values or return None if unable to supply the property.
 		"""
 		raise NotImplementedError
 
-	def GetPropValue(self, pIDString, dwIDStringLen, idProp):
+	def _hasProp(
+			self,
+			value: AcceptedGetPropTypes
+	) -> Optional[Tuple[BOOL, AcceptedGetPropTypes]]:
+		"""Constructs a tuple for the `IAccPropServer::GetPropValue` method, two elements:
+			1. `VARIANT pvarValue`
+			2. `BOOL pfHasProp` (either self.HAS_PROP or self.DOES_NOT_HAVE_PROP)"""
+		return value, self.HAS_PROP
+
+	def GetPropValue(
+			self, this, # unused "this" used to indicate to comTypes we want a low level implementation
+			pIDString: str,
+			dwIDStringLen: int,
+			idProp: GUID,
+			pvarValue: POINTER(VARIANT),
+			pfGotProp: POINTER(BOOL)
+	) -> int:
+		""" Exposed method to get a prop value.
+		see L{_getPropValue} for more details of args.
+		Uses a low-level approach, because comtypes tries to clear the VARIANT even though it is an out param.
+		When the pfHasProp part is FALSE / self.DOES_NOT_HAVE_PROP, then the pvarValue.vt part must be VT_EMPTY.
+		"""
+		# ensure exceptions don't leave this function. They will get get swallowed by the caller.
+		# instead catch and log exceptions.
 		try:
-			return self._getPropValue(pIDString, dwIDStringLen, idProp)
-		except Exception:
+			# Preset values for "no prop value", in case we return early.
+			pfGotProp.contents.value = self.DOES_NOT_HAVE_PROP
+			pvarValue.contents.vt = VT_EMPTY
+
+			ret = self._getPropValue(pIDString, dwIDStringLen, idProp)
+			if ret is None:
+				# We don't have the prop value, return early.
+				return S_OK
+			elif len(ret) != 2:
+				# We don't have the prop value, internal error.
+				raise RuntimeError("_getPropValue implementation must return None or two element tuple")
+			elif ret[1] != self.HAS_PROP:
+				# We don't have the prop value, return early.
+				return S_OK
+
+			# we do have the prop value
+			pfGotProp.contents.value = self.HAS_PROP
+			pvarValue.contents.value = ret[0]
+		except Exception as e: # catch and log all exceptions so they are not swallowed by caller.
 			log.exception()
```

Comment:
I think I still prefer returning something else than S_OK here. May be S_FALSE is a good one here, or comtypes.hresult.E_FAIL? Not sure how hresult codes work.

Reply:
I'm not so sure, I don't think we should be reporting an error unless it is the callers fault. If we would succeed with another value, then this essentially translates to a "look-up error", that outcome is handled by the `pfGotProp` outparam.

Reply:
Ah, fair enough.